analyze_map.go

```diff
@@ -0,0 +1,89 @@
+/**
+ * Copyright (c) F5, Inc.
+ *
+ * This source code is licensed under the Apache License, Version 2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+package crossplane
+
+import "fmt"
+
+// mapParameterMasks holds bit masks that define the behavior of the body of map-like directives.
+// Some map directives have "special parameter" with different behaviors than the default.
+type mapParameterMasks struct {
+	specialParameterMasks map[string]uint
+	defaultMasks          uint
+}
+
+//nolint:gochecknoglobals
+var mapBodies = map[string]mapParameterMasks{
+	"charset_map": {
+		defaultMasks: ngxConfTake1,
+	},
+	"geo": {
+		specialParameterMasks: map[string]uint{"ranges": ngxConfNoArgs, "proxy_recursive": ngxConfNoArgs},
+		defaultMasks:          ngxConfTake1,
+	},
+	"map": {
+		specialParameterMasks: map[string]uint{"volatile": ngxConfNoArgs, "hostnames": ngxConfNoArgs},
+		defaultMasks:          ngxConfTake1,
+	},
+	"match": {
+		defaultMasks: ngxConf1More,
+	},
+}
+
+// analyzeMapBody validates the body of a map-like directive. Map-like directives are block directives
+// that don't contain nginx directives, and therefore cannot be analyzed in the same way as other blocks.
+// nolint:gocyclo,gocognit
+func analyzeMapBody(fname string, parameter *Directive, term string, mapCtx string) error {
+	masks, known := mapBodies[mapCtx]
+
+	// if we're not inside a known map-like directive, don't bother analyzing
+	if !known {
+		return nil
+	}
+
+	if term != ";" {
+		return &ParseError{
+			What: fmt.Sprintf(`unexpected "%s"`, term),
+			File: &fname,
+			Line: &parameter.Line,
+		}
+	}
+
+	if mask, ok := masks.specialParameterMasks[parameter.Directive]; ok {
+		// use mask to check the parameter's arguments
+		if ((mask>>len(parameter.Args)&1) != 0 && len(parameter.Args) <= 7) || // NOARGS to TAKE7
```

Comment:
This same if statement is used twice in this function. Could you move the if condition into its own function to remove the code duplication? Doing this would also reduce the cognitive complexity of the analyzeMapBody function meaning that you would be able to remove the comment `nolint:gocyclo,gocognit` at the top of the function.

Reply:
Moved the condition to the new`hasValidArguments` helper and removed the `nolint` comment.